core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -1192,11 +1195,69 @@ object Chunk {
 
     override def apply(n: Int): A = vector(n)
 
-    override def foreach(f: A => Unit): Unit = vector.foreach(f)
+    override def foreach(f: A => Any): Unit = vector.foreach(f)
 
     override def toArray[A1 >: A](n: Int, dest: Array[A1]): Unit = { val _ = vector.copyToArray(dest, n, length) }
   }
 
+  final class BitChunk private (private val bytes: Chunk[Byte]) extends NonEmpty[Boolean] { self =>
+
+    def apply(n: Int): Boolean =
+      (bytes(n >> 3) & (1 << (7 - (n & 7)))) != 0
+
+    def foreach(f: Boolean => Any): Unit = {
+      var i = 0
+      while (i < length) {
+        f(apply(i))
+        i += 1
+      }
+    }
+
+    val length: Int =
+      bytes.length << 3
+
+    override def mkString: String =
+      mkString(", ")
+
+    override def mkString(sep: String): String =
+      mkString("BitChunk(", sep, ")")
+
+    override def mkString(start: String, sep: String, end: String): String = {
+      val builder = new scala.collection.mutable.StringBuilder
+      builder.append(start)
+      var i = 0
+      while (i < length) {
+        if (i != 0) {
+          builder.append(sep)
+        }
+        if (apply(i)) builder.append("1") else builder.append("0")
+        i += 1
+      }
+      builder.append(end)
+      builder.toString
+    }
+
+    def toArray[A1 >: Boolean](n: Int, dest: Array[A1]): Unit = {
+      var i = n
+      while (i < length) {
+        dest(i + n) = apply(i)
+        i += 1
+      }
+    }
+
+    def toBinaryString: String =
+      mkString("", "", "")
+
+    override def toString: String =
+      mkString
+  }
+
+  object BitChunk {
+
+    def fromByteChunk(bytes: Chunk[Byte]): BitChunk =
```

Comment:
We can add this to `Chunk`, as `Chunk#asBits`, when there exists implicit evidence that `A <:< Byte`.

Reply:
One thing we're running a bit into here is the new `NonEmptyChunk` concept. Right now `BitChunk` is extending `NonEmpty` which I think technically means we shouldn't be able to convert a `Chunk[Byte]` to a `BitChunk` unless we know statically that it is nonempty (or have to return an `Option` to express the fact that it might not be). But it seems like that is the last thing we want to do in performance sensitive code.

Reply:
I think it's ok to make that return `Option`, because the `Some` allocation overhead is amortized over the length of the bit chunk, which is hopefully quite large.

Reply:
Okay great. We may also have to think about the contract around emptiness with operations. If we want to allow creating subsets and still return a `BitChunk` then I I think we would have to accept the possibility that there could be an empty `BitChunk`.

Reply:
Hmmm. Would it be better for `BitChunk` to _not_ extend `NonEmpty`?

Reply:
I think it would be better for `BitChunk`. But then you kind of have an implementation leak where any other code within `Chunk` that wants to deal with empty versus nonempty chunks now has three cases to deal with. A chunk isn't either empty or nonempty. It is empty, nonempty, or this other bit chunk thing that could be one or the other.

Reply:
Ok, how about this: `BitChunk` extends `NonEmpty`, but in any case where we don't know if it's empty, we return only `Chunk`, and force the use to pattern match (or test) if they care about non-emptiness.

In this world, we would guard the constructor to `BitChunk` to make sure that it can only be invoked if there is at least one bit in it.

Reply:
I think that generally works. It may depend on how often users are going to want to do operations that reduce the number of bytes. If you want to do a bunch of those and lose type information every time so you can't chain operations it could get annoying.

Reply:
That's true, although even if you don't know it's a bit chunk, the bit chunk operations will still be optimized, so e.g. `bitChunk.take(5).take(3)` will still give you a BitChunk, assuming there were enough bits. So you could check at the end of a long series of operations, if you needed access to bit chunk specific methods.

Furthermore, it's possible the "bit chunk specific methods" could be exposed on chunk with evidence that `A <:< Boolean`, in which case they're not bit chunk specific anymore (just optimized quite nicely on bit chunks!), but rather boolean specific, so as you manipulate the size, you still get back something on which you can apply the bit chunk specific methods.

Reply:
It would be nice if we could do that because there is also a little weirdness here that right now you aren't allowed to know anything about the internal implementation of chunk, just whether or not it is empty, and now we're exposing this one really specialized case. So if we could pull it back to methods assuming a chunk is of a certain type that are then optimized under the hood would be very nice.

Reply:
Sounds good to me, let's try to make it work!

Reply:
Minimal implementation in the current version. Seems to work well and definitely nicer for the Chunk API.  The main downside I see is that it is easy to accidentally do operations like mapping over a chunk that turn it back into a chunk backed by an array.

Reply:
(1) If we override all methods in BitChunk, we can make sure that once you have a bit chunk, you stay with a bit chunk.

(2) Now making sure that if you start with something else, and end up with Chunk[Boolean], that it is backed by BitChunk, that would be harder, because it would mean that in any case where we create an `Array[Boolean]`, we need to actually change the logic and construct an `Array[Byte]` and set the bits inside there appropriately (maybe with a facade, a `BitChunkBuilder` that wraps an `Array[Byte]` and lets you change bits before building).

We don't have to do any of this work in the PR of course. We can split out tickets for (1) and (2) separately.